Select thumbnail, spee.ch upload

daovist opened the pull request:
This PR...

Lets `<FileSelector>` take custom labels

Adds `thumbnail_upload_statuses` constants

Adds `<SelectThumbnail>` component and uses it in `<PublishForm>`

Adds `ModalConfirmThumbnailUpload` which is called by `<SelectThumbnail>`

Adds `uploadThumbnailStatus` to `PublishState` (and not `PublishParams`) in `publishReducer`

Uses existing reducers and selectors, adds `doResetThumbnailStatus` and `doUploadThumbnail` actions for spee.ch requests

Adds `a.link` CSS

Cleans up some errors I got committing changes

daovist commented:
@liamcardenas this is ready for review. I think I got the rebase right but I'm not positive. I tested it post merge, commit, push, and it works fine.

A couple things that could/maybe should be done:

The modal and publish form write to the same `state.publish.nsfw` so if the thumbnail is marked nsfw it will be checked when the user gets to that option on the page. Should we turn off the ability to not mark a file nsfw when the thumbnail being uploaded was marked nsfw, by disabling the checkbox?

Should we disable publishing until the upload is complete? Until it does, the thumbnail field is a blank string, it doesn't get back the url with a claim_id until the upload is complete.

daovist commented:
@seanyesmunt this is working and ready for review

skhameneh commented on src/renderer/component/common/file-selector.jsx:
```diff
@@ -47,15 +49,14 @@ class FileSelector extends React.PureComponent<Props> {
+    const { type, currentPath, fileLabel, directoryLabel } = this.props;
+
+    const label =
+      type === 'file' ? fileLabel || __('Choose File') : directoryLabel || __('Choose Directory');
```
Can you add parens for legibility?

skhameneh commented on src/renderer/component/publishForm/view.jsx:
```diff
@@ -73,6 +77,10 @@ class PublishForm extends React.PureComponent<Props> {
     (this: any).getNewUri = this.getNewUri.bind(this);
   }
 
+  componentWillMount() {
```
This method is being deprecated, it's good to start using `getDerivedStateFromProps()`

https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

skhameneh commented on src/renderer/redux/actions/publish.js:
```diff
@@ -24,6 +34,82 @@ export const doUpdatePublishForm = (publishFormValue: UpdatePublishFormData) =>
+  const makeid = () => {
+    let text = '';
+    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+    for (let i = 0; i < 24; i += 1) text += possible.charAt(Math.floor(Math.random() * 62));
```
Can you split this line out with brackets for legibility?

skhameneh commented on src/renderer/redux/actions/publish.js:
```diff
@@ -24,6 +34,82 @@ export const doUpdatePublishForm = (publishFormValue: UpdatePublishFormData) =>
+  const data = new FormData();
+  const name = makeid();
+  const blob = new Blob([thumbnail], { type: `image/${fileExt.slice(1)}` });
+  data.append('name', name);
```
Can you clean up some of the formatting from ln90 to 110?
Line breaks, indentation, etc

daovist commented:
I cleaned this up based on comments from @skhameneh and other changes since the last time I worked on `select-thumbnail`. I moved constants to `lbry-redux` in [/lbryio/lbry-redux/pull/39](http://github.com/lbryio/lbry-redux/pull/39) which is a prereq for this PR.

I did not implement [getDerivedStateFromProps](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) thinking that pattern is all over the app and should be its own issue to clean up everywhere.

It now resets the thumbnail upload status when the whole form is reset. I refactored the errors and tested it on a faulty URL and everything worked like it should. Though some more @tzarebczan testing is probably in order.

seanyesmunt commented on src/renderer/constants/modal_types.js:
```diff
@@ -0,0 +1,23 @@
+export const SEND_TIP = 'send_tip';
+export const PUBLISH = 'publish';
+export const SEARCH = 'search';
+export const CONFIRM_THUMBNAIL_UPLOAD = 'confirmThumbnailUpload';
```
This should be in lbry-redux

seanyesmunt commented on src/renderer/redux/actions/publish.js:
```diff
@@ -30,6 +36,85 @@ export const doUpdatePublishForm = (publishFormValue: UpdatePublishFormData) =>
   });
 
+export const doResetThumbnailStatus = () => (dispatch: Dispatch): Action =>
+  fetch('https://spee.ch/api/channel/availability/@testing')
```
Is this the correct path?
